util/random.go

```diff
@@ -1,17 +1,56 @@
 package util
 
-import "math/rand"
+import (
+	cryptoRand "crypto/rand"
+	mathRand "math/rand"
+)
 
 // RandomString generates a random string of length n. Before calling this function, you should call
 // rand.Seed() to initialize the default source.
 //
 // Found at https://ispycode.com/Blog/golang/2016-10/How-to-generate-a-random-string-of-a-fixed-length
+//  - Adapted for more characters
 func RandomString(n int) string {
-	var letterRunes = []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ")
+	var letterRunes = []rune("0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz")
```

Comment:
I'm not sure there is any benefit in randomized order ... and perhaps quite the opposite, as sequential order makes it easy to spot any weakness in the runes. For instance, imagine there's duplicated chars, which would result in less evenly distributed function: it would be very hard to notice with a pre-randomized order.

Reply:
my concern here was that any internal randomize generator that messes with this "letters" (if its algorithm results can be predetermined) might make it more easy to reverse engineer the string if the reverse-engineer knows (or suspects) the seed was an ASCII-sequential string (of course, they could hunt down this line of code and see the actual string, so this is limited).

If you feel keeping an ASCII-sequential string that gets passed in is better, feel free to keep it. 

But in my opinion, at minimum, we should be using special characters in here too (not just alphanumeric chars) - thus increasing the span of possible characters used in the randomization process.

Reply:
Actually I realize my comment applies for `CryptoRandomString` below, and not this `RandomString`. I wonder if we really have to keep `RandomString` at all?

Talking only about `CryptoRandomString`: the runes are not a seed, they don't impact the underlying random bytes generation which is (supposed to be) secure - I agree that adding more characters (and perhaps increasing size of the output string, currently it's 15) would increase randomness ; but I still don't think we have to feel worried about the initial order, perhaps I'm wrong.

But yes, +1 for adding a couple of special characters, it doesn't cost anything.

Reply:
Citing the spec (https://openid.net/specs/openid-connect-core-1_0.html#NonceNotes):

> The nonce parameter value needs to include per-session state and be unguessable to attackers. One method to achieve this for Web Server Clients is to store a cryptographically random value as an HttpOnly session cookie and use a cryptographic hash of the value as the nonce parameter. In that case, the nonce in the returned ID Token is compared to the hash of the session cookie to detect ID Token replay by third parties. A related method applicable to JavaScript Clients is to store the cryptographically random value in HTML5 local storage and use a cryptographic hash of this value.

The important statement is that this string must be **unguessable to attackers**. As @jotak says, that sequential string is not a seed. A random string of 15 chars is being generated using the `crypto/rand` package which is supposed to be "cryptographycally secure" (see https://golang.org/pkg/crypto/rand/). So, I agree that it's not a problem to have that sequential string. I also don't see the need to add special characters (it's not a password :-) because the only requirement is to be unguessable. But, I'll anyway add a few chars.

What I see important in the `implementation notes` of the spec is the statement that there are two parts: (1) the random string, and (2) the cryptographic hash of that random string. The string (1) should live in the cookie, while the string (2) is the one passed to the IdP. Indeed, this makes sense to effectively block/mitigate replay attacks. The code, as it is, is using (1) for both and this doesn't result in a right mitigation (if you steal an `id_token`, you can craft the cookie to authenticate into Kiali) -- surprisingly, lots of examples in the internet suggest using the random string (1) for both the cookie and sending it to the IdP :worried: 

I'll check how to generate a `hash` to send to the IdP. It is passed in the URL, so there is a limitation in size. Once this is in place, I'm not sure if it's really useful to add special characters, but I'll add them anyway.

Reply:
I added the special characters and also did the change about sending a hash to the IdP instead of the plain nonce code.